### PR TITLE
feat(react): adding ability to support theming of datahub, with two themes included OOB

### PR DESCRIPTION
**Pull request by gabe-lyons**
Light theme:
image

image

image

Dark theme:
image
image
image

Instructions on how to use in the readme

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

**Comment by shirshanka** [comment]
@gabe-lyons : looks like there is a build issue

File: datahub-web-react/src/conf/theme/theme_light.config.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        "@text-color-secondary": "fade(black, 45%)",
+        "@heading-color": "fade(black, 85%)",
+        "@background-color-light": "hsv(0, 0, 98%)",
+        "@divider-color": "fade(black, 6%)",
+        "@disabled-color": "fade(black, 25%)",
+        "@steps-nav-arrow-color": "fade(black, 25%)",
+        "@homepage-background-upper-fade": "#132935",
+        "@homepage-background-lower-fade": "#FFFFFF"
+    },
+    "assets": {
```
Comment (jjoyce0510) [inline datahub-web-react/src/conf/theme/theme_light.config.json:20]: I like the centralization!! Is there a place we should document what values we support? Maybe in another PR?

Comment (jjoyce0510) [reply datahub-web-react/src/conf/theme/theme_light.config.json:20]: Oops -- looks like we have it below -- awesome!

File: datahub-web-react/src/conf/theme/theme_light.config.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
+{
+    "styles": {
+        "@border-radius-base": "8px",
```
Comment (jjoyce0510) [inline datahub-web-react/src/conf/theme/theme_light.config.json:3]: Why do these need the @ thing?

Comment (gabe-lyons) [reply datahub-web-react/src/conf/theme/theme_light.config.json:3]: We don't need it here, but this is the format of less style variables that antd uses:

https://github.com/ant-design/ant-design/blob/master/components/style/themes/default.less

When we merge them with the antd less variables we need the `@` signs. Whether we add theme here or add them in craco.config.js (https://github.com/linkedin/datahub/pull/2193/files#diff-fabf82065057e31dc626906f0279bc0ca3b309c3d964a9c4d5b5ff25f5beb5b0R11), either would be the same

File: datahub-web-react/src/App.tsx [hunk 3]
```diff
@@ -43,6 +47,14 @@ const client = new ApolloClient({
 });
 
 const App: React.VFC = () => {
+    const [dynamicThemeConfig, setDynamicThemeConfig] = useState<Theme | null>(null);
```
Comment (jjoyce0510) [inline datahub-web-react/src/App.tsx:50]: Why not just set this to default?

Comment (jjoyce0510) [reply datahub-web-react/src/App.tsx:50]: I see - is the idea that you can specify a "partial" theme that just overrides the default? This makes sense

File: datahub-web-react/src/App.tsx [hunk 4]
```diff
@@ -52,21 +64,38 @@ const App: React.VFC = () => {
         register.register(new TagEntity());
         return register;
     }, []);
+
+    const theme: Theme = useMemo(() => {
+        const overridesWithoutPrefix: { [key: string]: any } = {};
+        const themeConfig = dynamicThemeConfig || defaultThemeConfig;
+        Object.assign(overridesWithoutPrefix, themeConfig.styles);
+        Object.keys(overridesWithoutPrefix).forEach((key) => {
+            overridesWithoutPrefix[key.substring(1)] = overridesWithoutPrefix[key];
```
Comment (jjoyce0510) [inline datahub-web-react/src/App.tsx:73]: Why do we remove the "@" here... what if people don't use the @ when they specify styles?

File: datahub-web-react/src/App.tsx [hunk 5]
```diff
@@ -52,21 +64,38 @@ const App: React.VFC = () => {
...
-                        <Routes />
-                    </MockedProvider>
-                ) : (
-                    <ApolloProvider client={client}>
-                        <Routes />
-                    </ApolloProvider>
-                )}
-            </EntityRegistryContext.Provider>
-        </Router>
+        <ThemeProvider theme={theme}>
```
Comment (jjoyce0510) [inline datahub-web-react/src/App.tsx:83]: Awesome!

File: datahub-web-react/src/app/home/HomePageBody.tsx [hunk 6]
```diff
@@ -1,32 +1,40 @@
...
+        font-size: 32px;
+        color: ${(props) => props.theme.styles['homepage-background-upper-fade']};
+    }
+`;
+
+const EntityGridRow = styled(Row)`
+    padding: 40px 100px;
+`;
+
+const BodyContainer = styled.div`
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/home/HomePageBody.tsx:20]: this is great!

File: datahub-web-react/src/conf/theme/types.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
+export type Theme = {
+    styles: {
+        'border-radius-base': string;
```
Comment (jjoyce0510) [inline datahub-web-react/src/conf/theme/types.ts:3]: Why don't these require the @ guy?

**Review by jjoyce0510 — COMMENTED** [review]
This is awesome!!!! Great work @gabe-lyons.

**Pull request merged**